src/models/forecasts.jl

```diff
@@ -4,48 +4,99 @@ const Forecasts = Vector{<:Forecast}
 # This is deprecated because only the legacy System uses it. Parsing code need to change to
 # build the new format.
 const SystemForecastsDeprecated = Dict{Symbol, Forecasts}
-const IssueTime = NamedTuple{(:resolution, :initialtime),
-                             Tuple{Dates.Period, Dates.DateTime}}
-const SystemForecasts = Dict{IssueTime, Forecasts}
-
-# The default deserialization of SystemForecasts doesn't work for these reasons:
-# - IssueTime fails to serialize correctly.
-# - Forecasts are abstract types. We need to know what concrete type to create during
-#   deserialization.
-#
-# The code below converts the forecast data to a different format to enable deserialization.
-
-struct _SystemForecastArrayForJSON
-    resolution::Dates.Period
-    initialtime::Dates.DateTime
-    forecasts::Forecasts
-    forecast_types::Vector{String}  # Encode the exact type so that deserialization doesn't have
-                            # to infer what it is.
+
+struct _ForecastKey
+    initial_time::Dates.DateTime
+    forecast_type::DataType
 end
 
-struct _SystemForecastsForJSON
-    forecasts::Vector{_SystemForecastArrayForJSON}
+const _Forecasts = Dict{_ForecastKey, Vector{<:Forecast}}
+
+function _get_forecast_initial_times(data::_Forecasts)::Vector{Dates.DateTime}
+    initial_times = Set{Dates.DateTime}()
+    for key in keys(data)
+        push!(initial_times, key.initial_time)
+    end
+
+    return sort!(Vector{Dates.DateTime}(collect(initial_times)))
 end
 
-function JSON2.write(io::IO, forecasts::SystemForecasts)
-    return JSON2.write(io, encode_for_json(forecasts))
+function _add_forecast!(data::_Forecasts, forecast::T) where T <: Forecast
+    key = _ForecastKey(forecast.initialtime, T)
+    if !haskey(data, key)
+        data[key] = Vector{T}()
+    end
+
+    push!(data[key], forecast)
 end
 
-function JSON2.write(forecasts::SystemForecasts)
-    return JSON2.write(encode_for_json(forecasts))
+"""Container for forecasts and their metadata. Implementation detail that is not exported.
+Functions to access the data should go through the System."""
```

Comment:
Note this point and let me know if you disagree.

Reply:
I agree with this implementation.